dali/kernels/signal/resampling.h

```diff
@@ -0,0 +1,194 @@
+};
+
+void windowed_sinc(ResamplingWindow &window,
+    int coeffs, int lobes, std::function<double(double)> envelope = Hann) {
+  float scale = 2.0f * lobes / (coeffs - 1);
+  float scale_envelope = 2.0f / coeffs;
+  window.coeffs = coeffs;
+  window.lobes = lobes;
+  window.lookup.resize(coeffs + 2);  // add zeros
+  int center = (coeffs - 1) * 0.5f;
+  for (int i = 0; i < coeffs; i++) {
+    float x = (i - center) * scale;
+    float y = (i - center) * scale_envelope;
+    float w = sinc(x) * envelope(y);
+    window.lookup[i + 1] = w;
+  }
+  window.center = center + 1;  // allow for leading zero
+  window.scale = 1 / scale;
+}
+
+
+inline int64_t resampled_length(int64_t in_length, double in_rate, double out_rate) {
+  return std::ceil(in_length * out_rate / in_rate);
+}
+
+struct Resampler {
+  ResamplingWindow window;
+
+  void initialize(int lobes = 16, int lookup_size = 2048) {
+    windowed_sinc(window, lookup_size, lobes);
+  }
+
+  /**
+   * @brief Resample single-channel signal and convert to Out
+   *
+   * Calculates a range of resampled signal.
+   * The function can seamlessly resample the input and produce the result in chunks.
+   * To reuse memory and still simulate chunk processing, adjust the in/out pointers.
+   */
+  template <typename Out>
+  void resample(
+        Out *out, int64_t out_begin, int64_t out_end, double out_rate,
+        const float *in, int64_t n_in, double in_rate) const {
+    int64_t in_pos = 0;
+    int64_t block = 1 << 10;  // still leaves 13 significant bits for fractional part
+    double scale = in_rate / out_rate;
+    float fscale = scale;
+    for (int64_t out_block = out_begin; out_block < out_end; out_block += block) {
+      int64_t block_end = std::min(out_block + block, out_end);
+      double in_block_f = out_block * scale;
+      int64_t in_block_i = std::floor(in_block_f);
+      float in_pos = in_block_f - in_block_i;
+      const float *in_block_ptr = in + in_block_i;
+      for (int64_t out_pos = out_block; out_pos < block_end; out_pos++, in_pos += fscale) {
+        int i0, i1;
+        std::tie(i0, i1) = window.input_range(in_pos);
+        if (i0 + in_block_i < 0)
+          i0 = -in_block_i;
+        if (i1 + in_block_i >= n_in)
+          i1 = n_in - 1 - in_block_i;
+        float f = 0;
+        float x = i0 - in_pos;
+        for (int i = i0; i <= i1; i++, x++) {
+          assert(in_block_ptr + i >= in && in_block_ptr + i < in + n_in);
+          float w = window(x);
+          f += in_block_ptr[i] * w;
+        }
+        assert(out_pos >= out_begin && out_pos < out_end);
+        out[out_pos] = ConvertSatNorm<Out>(f);
+      }
+    }
+  }
+
+
+  /**
+   * @brief Resample multi-channel signal and convert to Out
+   *
+   * Calculates a range of resampled signal.
+   * The function can seamlessly resample the input and produce the result in chunks.
+   * To reuse memory and still simulate chunk processing, adjust the in/out pointers.
+   */
+  template <typename Out>
```

Comment:
Have you considered the static definition of a number of channels? Would it help with the perf?

Reply:
It would. I can do the VALUE_SWITCH trick here, no problem.